llvm/lib/Transforms/FuzzIntrospector/FuzzIntrospector.cpp

```diff
@@ -1122,13 +1200,239 @@ PreservedAnalyses FuzzIntrospectorPass::run(Module &M, ModuleAnalysisManager &AM
 // LLVM currently does not support dynamically loading LTO passes. Thus,
 // we dont register it as a pass as we have hardcoded it into Clang instead.
 // Ref: https://reviews.llvm.org/D77704
-static RegisterPass<FuzzIntrospector> X("fuzz-introspector", "FuzzIntrospector Pass",
-                                 false,
-                                 false );
+static RegisterPass<FuzzIntrospector> X("fuzz-introspector", "FuzzIntrospector
+Pass", false, false );
 
 static RegisterStandardPasses
     Y(PassManagerBuilder::EP_FullLinkTimeOptimizationEarly,
       [](const PassManagerBuilder &Builder, legacy::PassManagerBase &PM) {
         PM.add(new FuzzIntrospector());
       });
 */
+
+void FuzzIntrospector::branchProfiler(Module &M) {
+  std::string OutFileName = getNextLogFile() + ".branchProfile.yaml";
+  std::vector<BranchProfileEntry> OutMap;
+
+  logPrintf(L1, "We are in branch profiler.\n");
+
+  for (const auto &F : M) {
+    // Skip declarations or the functions that are not wrapped e.g. not
+    // reachable from entry point
+    if (F.isDeclaration() ||
+        FuncComplexityMap.find(&F) == FuncComplexityMap.end()) {
+      continue;
+    }
+    auto fName = F.getName().str();
+    logPrintf(L1, "We are in branch profiler for %s\n", fName.c_str());
+
+    // This map is function level
+    std::map<BasicBlock *, size_t> BBComplexityMap;
+
+    for (const auto &BB : F) {
+      auto TI = BB.getTerminator();
+      auto BI = dyn_cast<BranchInst>(TI);
+      if (BI && BI->isConditional()) {
+        auto Side0 = BI->getSuccessor(0);
+        auto Side1 = BI->getSuccessor(1);
+        auto BILoc = BI->getDebugLoc();
+
+        auto Reachable0 = findReachables(Side0);
+        auto Reachable1 = findReachables(Side1);
+
+        std::pair<size_t, size_t> Complexities =
+            findComplexities(Reachable0, Reachable1, BBComplexityMap);
+
+        auto Side0Comp = Complexities.first;
+        auto Side1Comp = Complexities.second;
+
+        std::pair<std::string, std::string> DbgExtracts;
+        DbgExtracts = getInsnDebugInfo((Instruction *)BI);
+        auto BRstring = DbgExtracts.first;
+        if (BRstring.length() == 0) {
+          continue; // Failed to get debug info
+        }
+        DbgExtracts = getBBDebugInfo(Side0, BILoc);
+        auto Side0String = DbgExtracts.first;
```

Comment:
In general for my liking there's a bit too much use of `auto`. The problem I have is when `auto` is used when the type is not explicitly clear from reading the code, e.g. in return values from functions that are not ubiquitous or fields in classes/structs that are not ubiquitous -- and I don't think we should use `auto` and then move the "String" to the variable name.

Reply:
Fair point. I plan to refactor this part of the code.